Add redelegation to the pseudocode model (#45)
informalsystems/partnership-heliax: 2023/Q1/artifacts/PoS-pseudocode/PoS-model-redelegation.md

# Namada's Cubic Proof-of-Stake

- [Namada's Cubic Proof-of-Stake](#namadas-cubic-proof-of-stake)
  - [Assumptions/simplifications](#assumptionssimplifications)
  - [Definitions](#definitions)
    - [Redelegations' Specific Terminology](#redelegations-specific-terminology)
  - [Technical Specification](#technical-specification)
    - [Data types](#data-types)
    - [Constants](#constants)
    - [Variables](#variables)
    - [Validator transactions](#validator-transactions)
    - [Bonding Mechanism Transactions](#bonding-mechanism-transactions)
    - [Main PoS functions](#main-pos-functions)
    - [Validator's functions](#validators-functions)
    - [Auxiliary functions](#auxiliary-functions)

## Assumptions/simplifications

- There is a single token type
- There is an initial set of active and inactive validators
- There are always enough non-jalied, candidate validators to fulfill the active set of validators at any given epoch

## Definitions

> - `pipeline_length` and `unbonding_length` define two epoch offsets. Note that `pipeline_length` must be greater or equal to `unbonding_length`.

Reviewer: ```suggestion
- `pipeline_length` and `unbonding_length` define two epoch offsets. Note that `pipeline_length` must be equal to or greater than `unbonding_length`.
```

### Redelegations' Specific Terminology

- A redelegation involves two validators. We use `source validator` to refer the validator from which the tokens are taken and `destination validator` to to refer to the validator to which the tokens are delegated.

- Given a `redelegation`, the `redelegation's starting epoch` is the epoch at which the redelegation transaction is processed. The `redelegation's ending epoch` is the epoch at which the redelegated tokens start contributing to the destination delegator. In the current design, if a redelegation's starting epoch is `e`, then its ending epoch would be `e + pipeline_length`. In the text throughout the specification, for a given `redelegation`, we use `redelegation.start` and `redelegation.end` to refer to the starting and ending epoch of the redelegation.

> - A redelegation is composed by one or more bonds (aka `redelegation bonds`). Each redelegation bond is a pair the epoch at which the redelegated tokens started contributing to the stake of the source validator and the amount of tokens. It is important to keep track of the starting epochs in order to apply slashes precisely.

Reviewer: ```suggestion
- A redelegation is composed of one or more bonds (aka `redelegation bonds`). Each redelegation bond is a pair of the epoch at which the redelegated tokens started contributing to the stake of the source validator and the amount of tokens. It is important to keep track of the starting epochs in order to apply slashes precisely.
```

> - Given a redelegation, we call `redelegation slashing window` the set of consecutive epoch in which the destination validator may be slashed due to the misbehave of the source validator.

Reviewer: ```suggestion
- Given a redelegation, we call `redelegation slashing window` the set of consecutive epochs in which the destination validator may be slashed due to the misbehaviour of the source validator.
```

  - In the current design, the redelegation slashing window of a redelegation spans from `redelegation.start - unbonding_length` up to `redelegation.end - 1`.
  - The window's lower bound if determined by the fact that when a redelegation is issued, we apply all slashes of the source validator already processed. This means any slash for an infraction committed at en epoch `< redelegation.start - unbonding_length`.

>   - The window's upper bound is determined by the epoch at which the redelegated tokens stop contributing to the stake to the source validator.

Reviewer: ```suggestion
  - The window's upper bound is determined by the epoch at which the redelegated tokens stop contributing to the stake of the source validator.
```

}
```

```go
func is_chained_redelegation(dest_validator_address, delegator_address)
  // avoid iteration over validators and do it over incoming_redelegations
  forall (val in validators) do
    var redelegation = validators[dest_validator_address].redelegations[val][delegator_address]
    if (redelegation.epoch >= 0 && redelegation.epoch + unbonding_length > cur_epoch) then
      return true
  return false
```

```go

> tx_redelegate(src_validator_address, dest_validator_address, delegator_address)

Reviewer 2: ```suggestion
func tx_redelegate(src_validator_address, dest_validator_address, delegator_address)
```

Author: I am not using `func` for txs for some reason (see the rest).

{
  // Disallow re-delegation if the source validator is frozen (similar to unbonding)
  var src_frozen = read_epoched_field(validators[src_validator_address].frozen, cur_epoch, false)
  if (is_validator(src_validator_address, cur_epoch+pipeline_length) && src_frozen == false) then
    // Check that `incoming_redelegations[delegator_address]` for `dest_validator_address` either don't exist
    // or if they do, they cannot be slashed anymore (`end + unbonding_length <= cur_epoch`)

>     if is_chained_redelegation(dest_validator_address, delegator_address) then

Reviewer 2: do we need to also check it for the src validator? i.e. `is_chained_redelegation(src_validator_address, delegator_address)`

Reviewer 2: as discussed this should rather be
```suggestion
    if is_chained_redelegation(src_validator_address, delegator_address) then
```

    var bonded_tokens = sum{amount | <start, amount> in delbonds}
    // Unbond the tokens from `src_validator_address` at pipeline offset, but 
    // without recording it in the `unbonds[delegator_address][src_validator_address]`
    var record_unbonds = false
    var amount_after_slashing = unbond(src_validator_address, delegator_address, bonded_tokens, record_unbonds)
    // add a bond in the `dest_validator_address` for the amount after slashing
    // done manually to avoid account transfer
    bonds[delegator_address][dest_validator_address].deltas[cur_epoch+pipeline_length] += amount_after_slashing
    redelegated_bonds[delegator_address][dest_validator_address][cur_epoch+pipeline_length] = Redelegation{validator: src_validator_address,
                                                                                                           amount: amount_after_slashing}
    //save the epoch at which the redelegation occurs to track chained redelegations
    validators[dest_validator_address].redelegations[src_validator_address][delegator_address] = IncomingRedelegation{epoch: cur_epoch,

>                                                                                                                       amoun: amount_after_slashing}

Reviewer 2: ```suggestion
                                                                                                                      amount: amount_after_slashing}
```

}
```

The transaction `tx_redelegate` creates a redelegation from source validator `src_validator_address` to `dest_validator_address` for any tokens that a delegator `delegator_address` has currently delegated to the source validator. The transactions takes the following steps:

1. The function first checks that the source validator is not frozen. This is to prevent unbonding at an epoch `e` from a validator that it is known to have behaved, but the slash has not been processed yet.
2. Then it checks that the redelegation is not a chain redelegation, in which case it prevents it by returning.
   - The design does not track chain redelegation precisely.

>    - Instead approximates it by preventing a validator to redelegate tokens belonging to a delegator if it has recently received a redelegation of tokens belonging to the same delegator.

Reviewer: Recently, i.e. within the unbonding period?

Author: Sort of. I have added some text to clarify, but the text in definitions about chain redelegation explains this better.

>    - The `inconming_redelegations` variable is used to implement this mechanism.

Reviewer: ```suggestion
   - The `incoming_redelegations` variable is used to implement this mechanism.
```

> 3. The function computes the total amount of tokens that the delegator has bonded at the source validator (`bonded_tokens`).

Reviewer: It should be possible to only redelegate a portion of the tokens that the delegator originally had delegated to the source validator, as well, right?

Author: It should be possible. Do we want to do it? I could look into it.

Reviewer: Yes, this would be nice (at some point, it's not urgent)

        var amount_unbonded = min{amount, remain}
        var redelegated_amount = 0
        if redelegation_record != ⊥ then
          redelegated_amount = redelegation_record.amount

        var amount_unbonded_after_slashing = 0
        //compute amount after slashing if some of the tokens do not come from redelegation
        if redelegated_amount < amount_unbonded then 
          // set of slashes that happened while the bond was contributing to the validator's stake
          var set_slashes = {s | s in slashes[validator_address] && start <= s.epoch }
          amount_unbonded_after_slashing = compute_amount_after_slashing(set_slashes, amount_unbonded - redelegated_amount)
        //compute amount after slashing  of the tokens coming from redelegation
        if redelegation_record != ⊥ then
          set_slashes = set_slashes \union {slash | slash in slashes[redelegation_record.validator] &&
                                                    //start - pipeline_length is the epoch when the redelegation was issued
                                                    //any slashed processed at an epoch < slash - pipeline_length - unbonding_length has been applied already

>                                                     start - pipeline_length - unbonding_length <= slash.epoch < start}

Reviewer 2: If I understand correctly, for a re-delegation we cannot determine the start epoch on the source validator, so we're just applying the slash indiscriminately as long as it's in the slashable duration here, right?

I don't think it's feasible to apply these exactly unless we do something even more complicated, as the re-delegation might be created from multiple bond deltas each with different start epoch, so I think it's is a good approximation, but would like to confirm and document this.

Reviewer 2: This ofc applies to the replicated logic in withdraw and end_of_epoch slashing

Author: > If I understand correctly, for a re-delegation we cannot determine the start epoch on the source validator, so we're just applying the slash indiscriminately as long as it's in the slashable duration here, right?

> I don't think it's feasible to apply these exactly unless we do something even more complicated, as the re-delegation might be created from multiple bond deltas each with different start epoch, so I think it's is a good approximation, but would like to confirm and document this.

We do not need to know when it started because we are slashing when doing the redelegation. Thus, we are not doing any approximation. Let me elaborate a bit:
- When the redelegation is issued, we compute the redelegated amount after slashing the source validator.
- If the redegelation happens at `e`, this means that we have processed all slashes up to `e-unbonding_lengh` (not included): an infraction at an epoch is processed at the end of that epoch + `unbonding_length`.
- Thus, when ubonding, we need to apply slashes for the source validator from `e-unbonding_lengh` to `e+pipeline_lentgh` (not included): when the redelegation ends.

Does this clarify it?

Reviewer 2: Sry, I'm not sure if I follow. Let me try to ask with 2 examples:

1. Say you have a delegation to validator A with start epoch `e`. In epoch `e+1` you redelegate to validator B. After that, there's evidence discovered for validator A from epoch `e-1`. Would the redelegation get slashed for it when its being processed at `e-1+unbonding` epoch? In this case, the original bond didn't contribute to the validator A voting power when the misbehavior occurred.
2. As another example, let's say that the delegation to validator A has start epoch `e`. Now you redelegate to B in `e+2`, after which an evidence for A is found for `e+1`. Does this differ from the first example when the slash is being processed at `e+1+unbonding` epoch? In this case, the original bond did contribute to the voting power of the validator when it misbehaved.

It seems to me that in both cases the slash would be applied to the redelegation.

Author: > 1. Say you have a delegation to validator A with start epoch `e`. In epoch `e+1` you redelegate to validator B. After that, there's evidence discovered for validator A from epoch `e-1`. Would the redelegation get slashed for it when its being processed at `e-1+unbonding` epoch? In this case, the original bond didn't contribute to the validator A voting power when the misbehavior occurred.

I see your concern. This is probably a problem: I believe in the current design we would slash. I am not sure there is a way to cut corners and avoid tracking the start of the delegation.

Reviewer 2: I think it’s an okay compromise. For redelegation there will be small chance that you may get slashed when you shouldn’t be, but not the other way around so it’s not a safety issue

Reviewer 2: Brought this up internally and we agreed we're okay with tracking the re-delegations with start epochs of their respective source bonds to make this precise. We're assuming epoch to last about a day and it's not gonna speed up significantly in future because of other concerns, so this shouldn't pose a DoS vector

Author: Addressed in [32102ae](https://github.com/informalsystems/partnership-heliax/pull/45/commits/32102aecfb854459ff15d81bc356898fcbba2d04).

          amount_unbonded_after_slashing += compute_amount_after_slashing(set_slashes, min{redelegated_amount, amount_unbonded})

        amount_after_slashing += amount_unbonded_after_slashing

        //update bond and create new unbond if record_unbonds=true
        bonds[delegator_address][validator_address].deltas[start] = amount - amount_unbonded

        var end = cur_epoch+pipeline_length+unbonding_length
        if record_unbonds then
          unbonds[delegator_address][validator_address].deltas[start, end] = amount_unbonded

        //manage redelegation records
        var redelegation_record_set_unbonds = ⊥
        if redelegation_record != ⊥ then
          
          if record_unbonds then
            update_unbond_redelegation_record(delegator_address,
                                              validator_address,
                                              redelegation_record.validator,
                                              start,
                                              end,

>                                               min{redelegation_record.amount, amount_unbonded})

Reviewer 2: I think we need to negate the amount here to subtract it from the record, right?

Reviewer 2: ah, nvm, this is the redelegation unbond

    //this serves to check that there are bonds (in the docs) and that these are greater than the amount we are trying to unbond
    if (sum{amount | <start, amount> in delbonds} >= total_amount) then
      var remain = total_amount
      var amount_after_slashing = 0
      //iterate over bonds and create unbonds
      forall (<start, amount> in delbonds while remain > 0) do

>         //Retrieve the bond's associated redelegation record (if any)

Reviewer: ```suggestion
        //Retrieve the bond's associated redelegation records (if any)
```

      update_validator_sets(validator_address, pipeline_length)
      return amount_after_slashing
}
```

```go
func update_unbond_redelegation_record(delegator_address, src_validator_address, dest_validator_address, start, end, amount)
{
  var unbond_redelegation_record = redelegated_unbonds[delegator_address][dest_validator_address][start, end]
  if (unbond_redelegation_record = ⊥) then
    var new_record = Redelegation{validator: src_validator_address, amount: amount}
  else
    var new_record = Redelegation{validator: src_validator_address, amount: unbond_redelegation_record.amount + amount}
  redelegated_unbonds[delegator_address][dest_validator_address][start, end] = new_record
}   
```

```go
/* COMMENT: Something to check for correctness https://github.com/informalsystems/partnership-heliax/pull/16#discussion_r924319213 */
//This function is called by transactions tx_withdraw_unbonds_validator and tx_withdraw_unbonds_delegator
func withdraw(validator_address, delegator_address)
{
  var delunbonds = {<start,end,amount> | amount = unbonds[delegator_address][validator_address].deltas[(start, end)] > 0 && end <= cur_epoch }
  //substract any pending slash before withdrawing

>   forall (<start,end,amount,redelegation_record> in selfunbonds) do

Reviewer 2: ```suggestion
  forall (<start,end,amount,redelegation_record> in delunbonds) do
```

}
```

```go
func compute_stake_fraction(infraction, voting_power){
  switch infraction
    case duplicate_vote: return duplicate_vote_rate * voting_power

>     case ligth_client_attack: return ligth_client_attack_rate * voting_power

Reviewer: ```suggestion
    case light_client_attack: return light_client_attack_rate * voting_power
```

    default: panic()
}
```

The function `end_of_epoch` is called at the end of every epoch. It takes the following steps:

1. It first computes the slash rate for any slash that has to be processed at the end of the current epoch.

> 2. The iterates over the enqueued slashes. For each slash:

Reviewer: ```suggestion
2. It then iterates over the enqueued slashes. For each slash:
```

```go
end_of_epoch()
{
  //iterate over all slashes for infractions within (-1,+1) epochs range (Step 2.1 of cubic slashing)
  var set_slashes = {s | s in enqueued_slashes[epoch] && cur_epoch-1 <= epoch <= cur_epoch+1}
  //calculate the slash rate (Step 2.2 of cubic slashing)
  var rate = compute_final_rate(set_slashes)

  var set_validators = {val | val = slash.validator && slash in enqueued_slashes[cur_epoch]}
  forall (validator_address in set_validators) do
    forall (slash in {s | s in enqueued_slashes[cur_epoch] && s.validator == validator_address}) do
      //set the slash on the now "finalized" slash amount in storage (Step 2.3 of cubic slashing)
      slash.rate = rate
      append(slashes[validator_address], slash)
      var total_staked = read_epoched_field(validators[validator_address].total_deltas, slash.epoch, 0)

>       slash_misbehaving_validator(validator_address, slash, staked_amount)

Reviewer 2: ```suggestion
      slash_misbehaving_validator(validator_address, slash, total_staked)
```

>       forall (val in validators) do

Reviewer 2: I think we'll want to avoid looping over all validators here to find the outgoing redelegations for `validator_address`. We could additionally record outgoing redelegations in `tx_redelegate` to make this a look-up without iteration.

To consider the DoS vector (https://github.com/informalsystems/partnership-heliax/pull/45#issuecomment-1519447279), we don't need to know the `delegator` address for these and so we can also group these by a triple of src validator, dest validator and redelegation epoch, e.g.:

```
outgoing_redelegations[src_validator_address][dest_validator_address][cur_epoch+pipeline_length] += amount_after_slashing
```

and then in here we should be able to iterate over `outgoing_redelegations[validator_address]` within the applicable epoch window

>         var pairs = { pair = <delegator, <epoch, redelegation> | pair in validators[var].redelegations[validator_address] &&

Reviewer 2: ```suggestion
        var pairs = { pair = <delegator, <epoch, redelegation> | pair in validators[val].redelegations[validator_address] &&
```

      var total_staked = read_epoched_field(validators[validator_address].total_deltas, slash.epoch, 0)
      slash_misbehaving_validator(validator_address, slash, total_staked)

      forall (dest_validators in validators[validator_address].outgoing_redelegations.keys()) do

>         forall (amount in {amount | amount = validators[validator_address].outgoing_redelegations[dest_validators][start] > 0 &&

Reviewer 2: I think there's a slight difference here from the previous version - the `start` epoch is not defined in this scope. We probably want to iter over the outgoing redelegations for each `validator_address, dest_validators` pair and use the same condition as before for their epochs, i.e. `redelegation.epoch - unbonding_length <= slash.epoch < redelegation.epoch + pipeline_length`

Author: That's the intention of the second condition in the line below: `start - unbonding_length <= slash.epoch < start + pipeline_length`. Maybe it is not well expressed though, we can always make the second iteration more explicit.

}
```

```go

> func slash_misbehaving_validator(src_validator_address, dest_validator_address, slash, amount)

Reviewer 2: ```suggestion
func slash_misbehaving_validator(src_validator_address, dest_validator_address, slash, total_staked)
```

      //retrieve redelegation record
      var redelegation_record = unbond.redelegation
      var redelegated_amount = 0
      if redelegation_record != ⊥ then
        redelegated_amount = redelegation_record.amount
      // set of slashes that happened while the bond was contributing to the validator's stake
      // We only need to apply a slash s if s.epoch < unbond.end - unbonding_length
      // It is easy to see that s.epoch + unbonding_length < slash.epoch implies s.epoch < unbond.end - unbonding_length
      // 1) slash.epoch = cur_epoch - unbonding_length
      // 2) unbond.end = cur_epoch + offset + unbonding_length => cur_epoch = unbond.end - offset - unbonding_length
      // By 1) s.epoch + unbonding_length < cur_epoch - unbonding_length
      // By 2) s.epoch + unbonding_length < unbond.end - offset - 2*unbonding_length => s.epoch < unbond.end - offset - 3*unbonding_length, as required.
      var set_slashes = {s | s in slashes[validator_address] &&
                             unbond.start <= s.epoch &&
                             s.epoch + unbonding_length < slash.epoch }

>       var total_unbonded = compute_amount_after_slashing(set_slashes, unbond.amount - redelegated_amount)

Reviewer 2: ```suggestion
     total_unbonded += compute_amount_after_slashing(set_slashes, unbond.amount - redelegated_amount)
```

                                              //any slashed processed at an epoch < unbond.start - pipeline_length has
                                              //been applied already, so we ignore them.
                                              //a slash with epoch e is processed at e + unbonding_length, we should then consider slashes
                                              //such that slash.epoch + unbonding_length >= unbond.start - pipeline_length
                                              unbond.start - pipeline_length - unbonding_length <= s.epoch < unbond.start &&
                                              //slash s is processed before the misbehaving epoch (slash.epoch)
                                              s.epoch + unbonding_length < slash.epoch}
        total_unbonded += compute_amount_after_slashing(set_slashes, redelegated_amount)

  var last_slash = 0
  // up to pipeline_length because there cannot be any unbond in a greater ß (cur_epoch+pipeline_length is the upper bound)
  forall (offset in 1..pipeline_length) do
    forall (unbond in validators[validator_address].set_unbonds[cur_epoch + offset] s.t. unbond.start <= slash.epoch) do

>     //retrieve redelegation record

Reviewer 2: I think the lines from here down to line 714 (`total_unbonded += ...`) are misaligned and should be inside the inner for loop

    update_voting_power(validator_address, offset)
}
```

```go
func slash_redelegated_validator(validator_address, src_validator, slash, staked_amount)
{
  var total_unbonded = 0
  //find the total unbonded from the slash epoch up to the current epoch first
  //a..b notation determines an integer range: all integers between a and b inclusive
  forall (epoch in slash.epoch+1..cur_epoch) do
    //filter those unbonds that were contributing to the stake when the validator misbehaved
    forall (unbond in validators[validator_address].set_unbonds[epoch] s.t. unbond.start - unbonding_length - pipeline_length <= slash.epoch && 
                                                                            unbond.redelegation != ⊥ &&
                                                                            unbond.redelegation.validator == src_validator) do
      //TODO: check that this set is not always empty
      var set_slashes = {s | s in slashes[validator_address] &&
                             unbond.start - unbonding_length - pipeline_length <= s.epoch &&
                             s.epoch + unbonding_length < slash.epoch }

>       total_unbonded = compute_amount_after_slashing(set_slashes, unbond.amount)

Reviewer 2: ```suggestion
      total_unbonded += compute_amount_after_slashing(set_slashes, unbond.amount)
```
